Enforce parsing and usage of existent roles

gguillemas opened the pull request:
Thank you for submitting this pull request. We really appreciate you spending the time to work on SurrealDB. 🚀 🎉 

## What is the motivation?

To address #5072 by enforcing that roles used in `DEFINE USER` exist and do not panic when being converted. 

## What does this change do?

Implements the `sql::Role` value, which directly relates to `iam::Role` in order to support role-specific parsing in SurrealQL. This `Role` value is used instead of `Ident` for storing user roles in the `DefineUserStatement` structure. This change allows removing the use of `unwrap` when converting a role from an `Ident` as a string, as well as identifying the use of a nonexistent role during parsing. A conversion function has been added to automatically migrate the `Vec<Ident>` field to `Vec<Role>`.

This change is made in the current context of roles being defined in code. The definition of the `Role` value will need to be updated when roles can be defined by the user, where having a specific `Role` value may still be useful.

**Note:** Queries that would previously parse successfully but could later result in a panic (e.g. `DEFINE USER example PASSWORD 'secret' ROLES manager`) will now fail to parse after merging this PR. The automated migration done by `revision` will simply ignore any nonexistent roles that were previously defined as part of a user definition. A warning will be printed on the server indicating this fact, allowing the user to overwrite those users with existent roles to remove the warning. This is preferable than forcing users to manage upgrade conflicts and the nonexistent roles would not have an effect on permissions anyhow.

## What is your testing strategy?

Ensure that existing tests pass and add new tests to verify that nonexistent roles are no longer accepted during parsing.

## Is this related to any issues?

Resolves #5072.

## Does this change need documentation?

No.

## Have you read the Contributing Guidelines?

<!-- All pull requests require that the contributing guidelines have been read and agreed to. -->

- [X] I have read the [Contributing Guidelines](https://github.com/surrealdb/surrealdb/blob/main/CONTRIBUTING.md)

emmanuel-keller reviewed (dismissed):
LGTM

gguillemas commented:
I am moving this back to draft because I have concerns regarding migrating a database that has some roles that cannot be automatically migrated by `revision` and will trigger an error. If some user definitions succeed in being converted to the new structure, an older version of SurrealDB will no longer be able to open the database to remove the users causing the issue.

I will look into simply removing any offending roles, as they previously did no have an effect anyhow.

gguillemas commented:
Finally, in order to prevent upgrade conflicts, nonexistent roles which cannot be converted to the new type will simply be ignored. This removes the possibility of conversion errors when accessing a database that had users already created with nonexistent roles with a new version of SurrealDB. See the PR description for a bit more detail.

garyhai commented:
> My concern is that using Role directly as a TokenKind might limit the extensibility of Role in future versions. Instead, I suggest using the internal iam::Role as TokenKind. This approach would provide more flexibility and make it easier to extend Role in the future.

Here is an example of the proposed change:

```rust
use crate::iam::Role;

enum TokenKind {
    // Other TokenKind variants
    Role(Role),
}
```

gguillemas commented:
> My concern is that using Role directly as a TokenKind might limit the extensibility of Role in future versions. Instead, I suggest using the internal iam::Role as TokenKind. This approach would provide more flexibility and make it easier to extend Role in the future.

Why could `iam::Role` be extended but not `sql::Role`? 🤔 

The idea behind wrapping types like we do with `sql::Algorithm` and others is to not expose types external crates (even our own) to the storage layer. I don't think there is any instance where we do that and I don't see a reason for doing that here. If we want to extend `sql::Role` we can always add a variant to the enum as we would with `iam::Role`, no?

garyhai commented:
I understand your design motivation. How about this approach: we keep the Role identifier type at the SQL layer, but constrain it during parsing to only allow the three predefined roles from iam::Role. My understanding is that iam::Role represents system roles, which correspond to a relatively stable set of actions and entities permissions. On the other hand, user-level Role is a combination of system roles or custom encapsulation of action and entity permissions sets, which can be very flexible. Therefore, my idea is to use iam::Role in TokenKind, extend iam::Role in sql::Role, and implement compatible conversion. For roles that exceed the current predefined roles, we can set the role identifier to None (Option) or Other (enum).

gguillemas commented:
> hich can be very flexible. Therefore, my idea is to use iam::Role in TokenKind, extend iam::Role in sql::Role, and implement compatible conversion. For roles that exceed the current predefined roles, we can set the role identifier to None (Option) or Other (enum).

I am sorry, but I don't think I understand what you are proposing exactly. Can you provide an example where doing this the way you describe would be beneficial? Can you explain why that would not be possible in the way that this PR implements it?

As implemented in this PR, `sql::Role` is exactly the same as `iam::Role` and provides the same flexibility, with some added flexibility due to the intermediate abstraction that allows for additional flexibility and specific differences between both. Checking for a hardcoded allowlist in the parser code is much worse (in my opinion) than having the enum itself be the thing to enforce the values during parsing, which also means the possible values are defined in `iam::Roles` and `sql::Roles` instead of the `define.rs` code where the parsing for user roles is implemented, which has nothing to do with IAM.

garyhai commented:
Thank you very much, @gguillemas, for the thorough explanation! I am still fairly new to SurrealDB and may have some misunderstandings about the overall design approach, so I appreciate your patience with any errors in my response.

It seems that our main point of difference might be in how we view the purpose of sql::Role. You mentioned, “That is exactly the benefit of having sql::Role as an intermediate abstraction of iam::Role,” whereas I had considered sql::Role as a composition and extension of iam::Role.

In my view, iam::Role provides a set of predefined, straightforward, and widely applicable roles. The sql::Role, then, could be used to combine and expand on these roles to create application-level roles, with names that are more intuitive to specific applications. In the future, finer-grained control over combinations of actions and entities could also allow users to define roles based on their business needs beyond the system-defined roles. For instance, some roles might only handle function writing, while others might focus exclusively on user management, and so forth.

With this in mind, I suggested that sql::Role might not serve as just an “intermediate” abstraction but as an extension of iam::Role through the use of enum types. When performing role checks, the current version could allow only roles defined in iam::Role, while other roles would trigger warnings but still default to the behavior of having no specific role settings (essentially, the default viewer permissions). Alternatively, we could introduce a special None role for any role lacking specific authorization.

Thank you again for the opportunity to discuss these ideas.

gguillemas commented:
> I appreciate your patience with any errors in my response

No worries.

> I had considered sql::Role as a composition and extension of iam::Role

Why an extension? It doesn't extend anything, it just [maps the variants of one enum to another](https://github.com/surrealdb/surrealdb/blob/6b6bee52c2df48b8d7f3c47000d496865dc4e04a/core/src/sql/role.rs)...

> The sql::Role, then, could be used to combine and expand on these roles to create application-level roles, with names that are more intuitive to specific applications. In the future, finer-grained control over combinations of actions and entities could also allow users to define roles based on their business needs beyond the system-defined roles.

This PR is under 200 lines and provides value in the present and the near future. Much like it was done, it can be undone do accommodate any changes that may require it. You are correct in us wanting to support user-defined roles in the future. We have a specific vision for it which will already require a significant overhaul of roles, including any code included in this PR. Rest assured that we are considering this and implementing `sql::Roles` as an abstraction will not impede these efforts in any way.

> With this in mind, I suggested that sql::Role might not serve as just an “intermediate” abstraction but as an extension of iam::Role through the use of enum types. 

What is it about enum types that you consider an extension even if they do not extend the enum? It's just mapping two types.

> Alternatively, we could introduce a special None role for any role lacking specific authorization.

Why do this when there is a perfectly good solution not involving this, which is the current method where the absence of roles defaults to no authorization? It seems like a hack to solve a problem that we do not have.

I believe I have heard and addressed your concerns. We would like to fix this bug and I would not like for this discussion to block our efforts to do so. We will always hear any concerns that our community has with our work, but I would like to consider this specific discussion settled.

gguillemas commented:
Thank you @garyhai. We will consider what you are proposing and weight it against the benefits of the changes as implemented. I am also mindful that this change may indeed be a breaking change to the Rust API, since roles will no longer be provided as an `Ident` but as a `Role`. I will take this to the rest of the SurrealDB team for discussion and come back with a final decision.

garyhai commented:
I understand now that my suggestions were somewhat off the mark due to my limited knowledge of SurrealDB’s future plans regarding Role. Given this, the fix you’ve proposed seems well-suited to the current needs, and I agree it should be implemented without delay.

Please consider my earlier suggestions as thoughts for me to reflect on further—I certainly don’t want them to hold up the process.

gguillemas commented:
> I understand now that my suggestions were somewhat off the mark due to my limited knowledge of SurrealDB’s future plans regarding Role. Given this, the fix you’ve proposed seems well-suited to the current needs, and I agree it should be implemented without delay.
> 
> Please consider my earlier suggestions as thoughts for me to reflect on further—I certainly don’t want them to hold up the process.

No worries. I appreciate you taking the time to review the changes and provide feedback. Ultimately, any additional scrutiny to the work that we do is welcome, especially with matters involving security. This PR will still need to be reviewed by an additional member of the SurrealDB team, who will take your comments into consideration.

gguillemas commented:
After further testing of this change and some discussion with the team, I am closing this PR in favor of #5092, which takes a simpler approach to resolving this issue. Despite our best efforts to use `revision` to make this change compatible between SurrealDB versions, we have found that SurrealDB clients containing this fix could have serialization issues with older servers.

In order to preserve the stability of SurrealDB, we have opted for a simpler fix that keeps structures intact but safely converts from `Ident` to `Role`, ensuring that this operation returns a descriptive error instead of panicking when handling nonexistent roles. This alternative fix also provides a simple and temporary check to ensure that only existent roles are accepted when defining a user.

Although we believe that the changes implemented in this PR are more in line with our development practices, we prefer to prioritize the stability of SurrealDB and minimizing potential inconvenience to its users.
